### PR TITLE
add meson

### PR DESCRIPTION
**Pull request by neheb**
Helps to avoid iconv problems that the CMake build has.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

### CODE REVIEW

File: meson.build [hunk 1]
```diff
@@ -0,0 +1,53 @@
+project(
+  'hidapi',
+  'c',
+  version : '0.12.0',
```
Comment (eli-schwartz) [inline meson.build:4]:
```suggestion
  version : files('VERSION'),
```

The other build systems read the version from here, might as well do it for Meson too. ;)

File: meson.build [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+incdirs = include_directories('hidapi')
+
+install_headers('hidapi/hidapi.h', subdir: 'hidapi')
+
+pconf = import('pkgconfig')
+
+if libusb_dep.found() and iconv_dep.found()
+  libhidapi = library('hidapi-libusb',
+    'libusb/hid.c',
+    dependencies: [ iconv_dep, libusb_dep ],
```
Comment (eli-schwartz) [inline meson.build:22]: This should depend on threads, I think.

Also, it is only called "-libusb" on linux.

And on Windows or macOS it uses the sources from windows/ or mac/ instead. See e.g. https://github.com/libusb/hidapi/blob/59e84ca74fae41a43f8ff3ea96b4f956c79c69b4/Makefile.am#L18-L40

File: meson.build [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    libhidapi,
+    description: 'C Library for USB HID device access from Linux, Mac OS X, FreeBSD, and Windows. This is the libusb implementation.',
+    url: 'https://github.com/libusb/hidapi',
+    subdirs: 'hidapi',
+  )
+endif
+
+if udev_dep.found()
+  hidraw_lib = library('hidapi-hidraw',
+    'linux/hid.c',
```
Comment (eli-schwartz) [inline meson.build:38]: I think this is only built on Linux.

**Review by eli-schwartz — CHANGES_REQUESTED** [review]

File: meson.build (hunk 2 above)
Comment (neheb) [reply meson.build:16]: threads probably implicitly gets included with libusb.

File: meson.build [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+install_headers('hidapi/hidapi.h', subdir: 'hidapi')
+
+pconf = import('pkgconfig')
+
+if libusb_dep.found() and iconv_dep.found()
+  libhidapi = library('hidapi-libusb',
+    'libusb/hid.c',
+    dependencies: [ iconv_dep, libusb_dep ],
+    include_directories: incdirs,
+    version: files('VERSION'),
```
Comment (eli-schwartz) [inline meson.build:18]: This works for project, specifically.

For libraries just use meson.project_version()...

File: meson_options.txt [hunk 5]
```diff
@@ -0,0 +1,7 @@
+option('hidraw', type : 'feature',
+  description : 'Build hidraw. Requires udev.',
```
Comment (JoergAtGithub) [inline meson_options.txt:2]:
```suggestion
description : 'Build hidapi with hidraw backend. Requires udev.',
```

File: meson_options.txt [hunk 6]
```diff
@@ -0,0 +1,7 @@
...
+)
+
+option('libusb', type : 'feature',
+  description : 'Build hidapi. Requires libusb.',
```
Comment (JoergAtGithub) [inline meson_options.txt:6]:
```suggestion
description : 'Build hidapi with libusb backend. Requires libusb.',
```

**Comment by Youw** [comment]
What does it solve compared to #410?

> avoid iconv problems that the CMake build has

There is 0 known issues related to `iconv`: https://github.com/libusb/hidapi/issues?q=is%3Aissue+is%3Aopen+iconv

Can you elaborate?

**Comment by neheb** [comment]
> What does it solve compared to #410?
> 
> > avoid iconv problems that the CMake build has
> 
> There is 0 known issues related to `iconv`: https://github.com/libusb/hidapi/issues?q=is%3Aissue+is%3Aopen+iconv
> 
> Can you elaborate?

didn't know there was another PR.

Anyway, issue is that when both external and internal iconv implementations exist, CMake's find_package(Iconv) fails to find iconv and terminates the build. More info here:

https://github.com/openwrt/packages/actions/runs/2477509891
https://github.com/openwrt/packages/pull/18738

solved by using dependency('iconv') which just works.

**Comment by Youw** [comment]
I'm pretty sure, we can find a solution to CMake's `find_package(Iconv)`.

---
I do not support having yet another alternative build system for HIDAPI when it can be avoided - you'd have to duplicate the entire build logic we already have in CMake.
Lets focuse on #410 instead.

**Pull request closed without merging**